plutarch-api/src/Plutarch/Api.hs

```diff
@@ -0,0 +1,1275 @@
+  type PTryFromExcess PData (PAsData PPosixTime) = Mret PPosixTime
+  ptryFrom' ::
+    forall (s :: S) (r :: S -> Type).
+    Term s PData ->
+    ((Term s (PAsData PPosixTime), Reduce (PTryFromExcess PData (PAsData PPosixTime) s)) -> Term s r) ->
+    Term s r
+  ptryFrom' opq = runTermCont $ do
+    (wrapped :: Term s (PAsData PInteger), unwrapped :: Term s PInteger) <-
+      tcont $ ptryFrom @(PAsData PInteger) opq
+    tcont $ \f -> pif (0 #<= unwrapped) (f ()) (ptraceError "ptryFrom(POSIXTime): must be positive")
+    pure (punsafeCoerce wrapped, pcon $ PPosixTime unwrapped)
+
+-- | @since 2.0.0
+newtype PInterval (a :: S -> Type) (s :: S)
+  = PInterval
+      ( Term
+          s
+          ( PDataRecord
+              '[ "from" ':= PLowerBound a
+               , "to" ':= PUpperBound a
+               ]
+          )
+      )
+  deriving stock
+    ( -- | @since 2.0.0
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      PlutusType
+    , -- | @since 2.0.0
+      PIsData
+    , -- | @since 2.0.0
+      PDataFields
+    , -- | @since 2.0.0
+      PEq
+    , -- | @since 2.0.0
+      PPartialOrd
+    , -- | @since 2.0.0
+      POrd
+    , -- | @since 2.0.0
+      PShow
+    )
+
+-- | @since 2.0.0
+instance DerivePlutusType (PInterval a) where
+  type DPTStrat _ = PlutusTypeData
+
+-- | @since 2.0.0
+instance
+  PLiftData a =>
+  PUnsafeLiftDecl (PInterval a)
+  where
+  type PLifted (PInterval a) = (Plutus.Interval (PLifted a))
+
+-- | @since 2.0.0
+deriving via
+  (DerivePConstantViaData (Plutus.Interval a) (PInterval (PConstanted a)))
+  instance
+    PConstantData a =>
+    PConstantDecl (Plutus.Interval a)
+
+-- | @since 2.0.0
+newtype PPubKeyHash (s :: S) = PPubKeyHash (Term s PByteString)
+  deriving stock
+    ( -- | @since 2.0.0
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      PlutusType
+    , -- | @since 2.0.0
+      PIsData
+    , -- | @since 2.0.0
+      PEq
+    , -- | @since 2.0.0
+      PPartialOrd
+    , -- | @since 2.0.0
+      POrd
+    , -- | @since 2.0.0
+      PShow
+    )
+
+-- | @since 2.0.0
+instance DerivePlutusType PPubKeyHash where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since 2.0.0
+instance PUnsafeLiftDecl PPubKeyHash where
+  type PLifted PPubKeyHash = Plutus.PubKeyHash
+
+-- | @since 2.0.0
+deriving via
+  (DerivePConstantViaBuiltin Plutus.PubKeyHash PPubKeyHash PByteString)
+  instance
+    PConstantDecl Plutus.PubKeyHash
+
+-- | @since 2.0.0
+instance PTryFrom PData (PAsData PPubKeyHash) where
+  type PTryFromExcess PData (PAsData PPubKeyHash) = Mret PPubKeyHash
+  ptryFrom' opq = runTermCont $ do
+    unwrapped <- tcont . plet $ ptryFrom @(PAsData PByteString) opq snd
+    tcont $ \f ->
+      pif
+        (plengthBS # unwrapped #== 28)
+        (f ())
+        (ptraceError "ptryFrom(PPubKeyHash): must be 28 bytes long")
+    pure (punsafeCoerce opq, pcon . PPubKeyHash $ unwrapped)
+
+-- | @since 2.0.0
+newtype PubKey = PubKey
```

Comment:
Why do we need this here? Feels a bit random and unrelated to Plutarch

Reply:
This is a direct port from how it was defined. It's because it acts as the Haskell-level equivalent to `PPubKey` I think.